src/clp_ffi_js/ir/ClpIrV1Decoder.cpp

```diff
@@ -0,0 +1,233 @@
+        );
+        throw ClpJsException(
+                clp::ErrorCode::ErrorCode_MetadataCorrupted,
+                __FILENAME__,
+                __LINE__,
+                "Failed to decompress"
+        );
+    }
+
+    return ClpIrV1Decoder(std::move(data_buffer), zstd_decompressor, std::move(result.value()));
+}
+
+auto ClpIrV1Decoder::get_estimated_num_events() const -> size_t {
+    return m_log_events.size();
+}
+
+auto ClpIrV1Decoder::build_idx(size_t begin_idx, size_t end_idx) -> emscripten::val {
+    constexpr size_t cFullRangeEndIdx{0};
+    if (0 != begin_idx && cFullRangeEndIdx != end_idx) {
+        throw ClpJsException(
+                clp::ErrorCode::ErrorCode_Unsupported,
+                __FILENAME__,
+                __LINE__,
+                "Partial range index building is not yet supported."
+        );
+    }
+    if (false == m_full_range_built) {
+        m_full_range_built = true;
+        constexpr size_t cDefaultNumLogEvents{500'000};
+        m_log_events.reserve(cDefaultNumLogEvents);
+        while (true) {
+            auto const result{m_deserializer.deserialize_log_event()};
+            if (false == result.has_error()) {
+                m_log_events.emplace_back(result.value());
+                continue;
+            }
+            auto const error{result.error()};
+            if (std::errc::no_message_available == error) {
+                break;
+            }
+            if (std::errc::result_out_of_range == error) {
+                SPDLOG_ERROR("File contains an incomplete IR stream");
+                break;
+            }
+            throw ClpJsException(
+                    clp::ErrorCode::ErrorCode_Corrupt,
+                    __FILENAME__,
+                    __LINE__,
+                    "Failed to decompress: "s + error.category().name() + ":" + error.message()
+            );
+        }
+        m_data_buffer.reset(nullptr);
+    }
+
+    auto results{emscripten::val::object()};
+    results.set("numValidEvents", m_log_events.size());
+    results.set("numInvalidEvents", 0);
+    return results;
+}
+
+auto ClpIrV1Decoder::decode(size_t begin_idx, size_t end_idx) const -> emscripten::val {
+    if (m_log_events.size() < end_idx || begin_idx >= end_idx) {
+        return emscripten::val::null();
+    }
+
+    size_t log_num{begin_idx + 1};
+    std::string message;
+    constexpr size_t cDefaultReservedMessageLength{512};
+    message.reserve(cDefaultReservedMessageLength);
+    auto const results{emscripten::val::array()};
+    std::span<clp::ir::LogEvent<clp::ir::four_byte_encoded_variable_t> const> const log_events_span(
+            m_log_events.begin() + static_cast<std::ptrdiff_t>(begin_idx),
+            m_log_events.begin() + static_cast<std::ptrdiff_t>(end_idx)
+    );
+    for (auto const& log_event : log_events_span) {
+        message.clear();
+
+        // TODO: Replace below handlers code by an OSS decoding method once it's added in the future
```

Comment:
That method has been added to the LogEvent class, so shall we address this TODO?